src/NEPTypes.jl

```diff
@@ -179,122 +193,121 @@ julia> compute_Mder(nep,1)-(A0+A1*exp(1))
-                    Fid=nep.fi[i](reshape([Sd[1]],1,1))[1]*ones(size(Sd,1))
+                    Fid=nep.fi[i](Sd[1])*ones(T0,p)
                 else  # Diagonal but not constant
-                    Fid=zeros(ComplexF64,size(S,1))
-                    for j=1:size(S,1)
-                        Fid[j]=nep.fi[i](reshape([Sd[j]],1,1))[1]
+                    Fid0=Vector{Number}(undef,p)
+                    for j=1:p
+                        Fid0[j]=nep.fi[i](Sd[j])
                     end
+                    Fid=Vector{FStype}(Fid0);
                 end
-                Fi=sparse(Diagonal(Fid))
-            else  # Otherwise just compute the matrix function operation
-                if(nep.Schur_factorize_before)
-                    Fi= Q*nep.fi[i](T)*Q'
-                else
-                    Fi=nep.fi[i](S)
-                end
-            end
-            ## Sum it up
-
-
-            VFi=V*Fi;
-            if (isa(nep.A[i],SubArray) && (eltype(nep.A[i]) != eltype(VFi)))
-                # 2018-09-14: SubArray x Matrix of different types do not work? Check in new version of Julia/base at later point (TODO)
-                # https://discourse.julialang.org/t/subarray-x-matrix-multipliciation-of-different-eltype-fails/14950
-                # Workaround by making a matrix copy.
-                Z=Z+copy(nep.A[i])*VFi;
+                Fi[:,:]=Diagonal(Fid)
             else
-                Z=Z+nep.A[i]*VFi;
+                Fi[:,:]=ff[i](S);
             end
-
+            mul!(VFi,V,Fi);
+            Z[:,:] .+= AA[i]*VFi;
         end
         return Z
     end
 
 
-    function compute_Mder(nep::SPMF_NEP,λ::Number)
-
-        # figure out the return type, as the greatest type of all input
-        x = map(i -> nep.fi[i](reshape([λ],1,1))[1], 1:length(nep.fi))
-        Tx = mapreduce(eltype, promote_type, x)
-        TA=mapreduce(eltype, promote_type, nep.A); # Greatest type of all A-matrices
-        TZ=promote_type(TA,Tx)  # output type
-        if isempty(nep.As)
-            # Full matrices
-            Z = zeros(TZ,size(nep,1),size(nep,1));
-            for k=1:size(nep.A,1)
-                Z += nep.A[k] * x[k]
-            end
-            return Z
-        else
-            Z = SparseMatrixCSC(nep.As[1].m, nep.As[1].n,
-                                nep.As[1].colptr, nep.As[1].rowval,
-                                convert.(TZ, nep.As[1].nzval .* x[1]))
-            for k = 2:length(nep.As)
-                Z.nzval .+= nep.As[k].nzval .* x[k]
-            end
+     function compute_Mder(nep::SPMF_NEP_dense,λ::Number)
+         # figure out the return type, as the greatest type of all input
+         x = map(i -> nep.fi[i](reshape([λ],1,1))[1], 1:length(nep.fi))
+         Tx = mapreduce(eltype, promote_type, x)
+         TA=mapreduce(eltype, promote_type, nep.A); # Greatest type of all A-matrices
+         TZ=promote_type(TA,Tx)  # output type
```

Comment:
Maybe we could extract the 5 lines above here, that are repeated below, to its own function? That would make it easier and safer to refactor that code in the future.

Reply:
Good idea. Done.